features/filesystem/bd/BlockDevice.h

```diff
@@ -0,0 +1,180 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2016 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef MBED_BLOCK_DEVICE_H
+#define MBED_BLOCK_DEVICE_H
+
+#include <stdint.h>
+
+
+/** Enum of standard error codes
+ *
+ *  @enum bd_error_t
+ */
+enum bd_error {
+    BD_ERROR_OK                 = 0,     /*!< no error */
+    BD_ERROR_WOULD_BLOCK        = -4001, /*!< operation would block */
+    BD_ERROR_UNSUPPORTED        = -4002, /*!< unsupported operation */
+    BD_ERROR_PARAMETER          = -4003, /*!< invalid parameter */
+    BD_ERROR_NO_INIT            = -4004, /*!< uninitialized */
+    BD_ERROR_NO_DEVICE          = -4005, /*!< device is missing or not connected */
+    BD_ERROR_WRITE_PROTECTED    = -4006, /*!< write protected */
+    BD_ERROR_DEVICE_ERROR       = -4007, /*!< device specific error */
+};
+
+/** Type representing either 0 or a negative error code
+ */
+typedef int32_t bd_error_t;
+
+/** Type representing the address of a specific block
+ */
+typedef uint64_t bd_addr_t;
+
+/** Type representing a quantity of 8-bit bytes
+ */
+typedef uint64_t bd_size_t;
+
+
+/** A hardware device capable of writing and reading blocks
+ */
+class BlockDevice {
+public:
+    /** Lifetime of a block device
+     */
+    virtual ~BlockDevice() {};
+
+    /** Initialize a block device
+     *
+     *  @return         0 on success or a negative error code on failure
+     */
+    virtual bd_error_t init() = 0;
+
+    /** Deinitialize a block device
+     *
+     *  @return         0 on success or a negative error code on failure
+     */
+    virtual bd_error_t deinit() = 0;
+
+    /** Read blocks from a block device
+     *
+     *  @param buffer   Buffer to write blocks to
+     *  @param addr     Address of block to begin reading from
+     *  @param size     Size to read in bytes, must be a multiple of read block size
+     *  @return         0 on success, negative error code on failure
```

Comment:
Added a little comment.

The only case in which read/write/erase/program return fewer than the number of bytes requested of the operation is when an error occurs. However, in that condition, an error code is returned.

Adding size to the return value adds extra work to the implementations and extra checks to the users that have no benefit from what I can tell.